src/ir/ssa_names.h

```diff
@@ -17,51 +17,89 @@
 #define SRC_IR_SSA_NAMES_H_
 
 #include "absl/container/flat_hash_map.h"
+#include "src/ir/value.h"
 
 namespace raksha::ir {
 
 class Operation;
 class Block;
 class Module;
+class Value;
 
 class SsaNames {
  public:
-  using ID = int;
-  SsaNames() {}
-  // Disable copy (and move) semantics.
+  using ID = std::string;
+  SsaNames()
+      : block_ids_("b"),
+        module_ids_("m"),
+        value_ids_("%"),
+        operation_ids_("%") {}
+
+  // Disable copy semantics.
   SsaNames(const SsaNames &) = delete;
   SsaNames &operator=(const SsaNames &) = delete;
+  SsaNames(SsaNames &&) = default;
 
   ID GetOrCreateID(const Operation &operation) {
     return operation_ids_.GetOrCreateID(&operation);
   }
 
+  ID UpdateID(const Operation &operation, ID id) {
+    return operation_ids_.GetOrCreateID(&operation, id);
+  }
   ID GetOrCreateID(const Block &block) {
     return block_ids_.GetOrCreateID(&block);
   }
 
+  ID UpdateID(const Block &block, ID id) {
+    return block_ids_.GetOrCreateID(&block, id);
+  }
+
   ID GetOrCreateID(const Module &module) {
-    return module_ids_.GetOrCreateID(std::addressof(module));
+    return module_ids_.GetOrCreateID(&module);
+  }
+  ID UpdateID(const Module &module, ID id) {
+    return module_ids_.GetOrCreateID(&module, id);
+  }
+
+  ID GetOrCreateID(const Value &value) {
+    return value_ids_.GetOrCreateID(value);
+  }
+
+  ID UpdateID(const Value &value, ID id) {
+    return value_ids_.GetOrCreateID(value, id);
   }
 
  private:
-  template <class T>
+  template <typename T>
   class IDManager {
    public:
-    ID GetOrCreateID(const T *entity) {
+    IDManager(absl::string_view prefix) : prefix_(prefix) {}
+
+    ID GetOrCreateID(T entity) {
+      // Note that if `entity` is already in the map, the `insert` call below
+      // will return the existing entity. Otherwise, a new entity is added.
+      auto insert_result =
+          item_ids_.insert({entity, absl::StrCat(prefix_, item_ids_.size())});
+      return insert_result.first->second;
+    }
+
+    ID GetOrCreateID(T entity, ID id) {
```

Comment:
Can you change this to `void ReplaceID` (or `void UpdateID`) and make sure it is changed even if the entity exists? You can use `insert_or_assign` to update the value. e.g.,

```
void ReplaceID(T entity, ID id) { item_ids_.insert_or_assign(entity, id);  }
```


Also, we will need to add tests.